### PR TITLE
ARROW-4236: [java] Distinct plasma client create exceptions

### PR DESCRIPTION
**Pull request by bibabolynn**
when ray puts an object in plasma store, there are 2 exceptions may be thrown, one is "An object with this ID already exists in the plasma store" and the other is "The plasma store ran out of memory and could not create this object",  distinct them rather than let them both be the same java class

@raulchen please help review

### CODE REVIEW

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (raulchen) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: `PlasmaOutOfMemoryException` doesn't take object id as the parameter.

File: java/plasma/src/main/java/org/apache/arrow/plasma/PlasmaClient.java
Comment (raulchen) [inline java/plasma/src/main/java/org/apache/arrow/plasma/PlasmaClient.java]: what is `threadConn`?

File: java/plasma/src/main/java/org/apache/arrow/plasma/PlasmaOutOfMemoryException.java
Comment (raulchen) [inline java/plasma/src/main/java/org/apache/arrow/plasma/PlasmaOutOfMemoryException.java]: remove `and could not create object with ID`

File: java/plasma/src/main/java/org/apache/arrow/plasma/PlasmaOutOfMemoryException.java
Comment (raulchen) [inline java/plasma/src/main/java/org/apache/arrow/plasma/PlasmaOutOfMemoryException.java]: new line

File: java/plasma/src/main/java/org/apache/arrow/plasma/DuplicateObjectException.java
Comment (raulchen) [inline java/plasma/src/main/java/org/apache/arrow/plasma/DuplicateObjectException.java]: new line

File: cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc
Comment (praveenbingo) [inline cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc]: indentation looks off.. also exception? for name..

File: java/plasma/src/main/java/org/apache/arrow/plasma/PlasmaClient.java
Comment (praveenbingo) [inline java/plasma/src/main/java/org/apache/arrow/plasma/PlasmaClient.java]: this looks backward incompatible..earlier we were ignoring any exceptions while creating the buffer

which is why the test is failing here - https://github.com/apache/arrow/blob/41a4e72f08021a682426ceb46c7921ad91a269cd/java/plasma/src/test/java/org/apache/arrow/plasma/PlasmaClientTest.java#L145

any reason to change the behavior?

Comment (raulchen) [reply java/plasma/src/main/java/org/apache/arrow/plasma/PlasmaClient.java]: The reason is because it makes more sense to propagate the exception to user code, and let user decide what to do. E.g., when putting an object with existing ID, someone may want to silently ignore the error, someone may want to use a different ID, and someone may want to terminate the program. The decision should depends on application logic.

**Comment by bibabolynn** [comment]
@praveenbingo  could you help check the ci result. I can't see why it failed or does it have anything to do with my changes. I've tried several times, but they all failed.

File: java/plasma/src/test/java/org/apache/arrow/plasma/PlasmaClientTest.java
Comment (praveenbingo) [inline java/plasma/src/test/java/org/apache/arrow/plasma/PlasmaClientTest.java]: do you also want to assert that an exception is thrown..

**Review by praveenbingo — COMMENTED** [review]
also looks like the python tests are having core dumps..would you be able to run the python tests to see if they are caused by this change 
(i am not an expert in plasma to comment authoritatively)

**Comment by wesm** [comment]
Can you please create a JIRA issue for this and update the PR title?

**Comment by guoyuhong** [comment]
https://issues.apache.org/jira/browse/ARROW-4236

**Comment by bibabolynn** [comment]
@wesm I've asked somebody else to create the JIRA issue and paste the link above. Because I don't have authority. 'bibabolynn' is my account, maybe you can add me

**Review by pcmoritz — APPROVED** [review]
The change looks good to me.

**Comment by pcmoritz** [comment]
@bibabolynn @raulchen Are there any changes necessary in Ray for this?

**Comment by pcmoritz** [comment]
> @wesm I've asked somebody else to create the JIRA issue and paste the link above. Because I don't have authority. 'bibabolynn' is my account, maybe you can add me

I added you as contributor and assigned the Issue to you.

**Comment by bibabolynn** [comment]
@wesm @pcmoritz @praveenbingo  since this pr has already been approved, I wonder when will it be merged?

**Comment by wesm** [comment]
@pcmoritz is also a committer so he can also merge patches. Can you please rebase to make sure the build is passing? We fixed a bunch of problems that were going on in CI a couple weeks ago

**Comment by praveenbingo** [comment]
Change looks good to me too..

Please rebase to re-trigger and verify the tests and this should be good for merge, as @wesm said @pcmoritz can merge the change post the tests..

**Comment by raulchen** [comment]
All tests are passing. @pcmoritz could you help merge it when you have time? Thanks!

**Comment by kou** [comment]
I re-run CI. Now, CI is green.

**Comment by pcmoritz** [comment]
Yes, merging it, thanks for the contribution (originally I left it open in case there are more comments).

**Pull request closed without merging**
